File: backend/app/api/routes/workspaces.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session
from typing import List
from app.core.database import get_session
from app.schemas.workspace import WorkspaceResponse, WorkspaceCreate, WorkspaceUpdate
from app.services import workspace_service

router = APIRouter()

@router.get("", response_model=List[WorkspaceResponse])
def list_workspaces(session: Session = Depends(get_session)):
    return workspace_service.get_workspaces(session)

@router.post("", response_model=WorkspaceResponse, status_code=status.HTTP_201_CREATED)
def create_workspace(workspace_in: WorkspaceCreate, session: Session = Depends(get_session)):
    return workspace_service.create_workspace(session, workspace_in)

@router.get("/{workspace_id}", response_model=WorkspaceResponse)
def get_workspace(workspace_id: str, session: Session = Depends(get_session)):
    workspace = workspace_service.get_workspace_by_id(session, workspace_id)
    if not workspace:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="Workspace not found")
    return workspace

@router.patch("/{workspace_id}", response_model=WorkspaceResponse)
def update_workspace(workspace_id: str, workspace_in: WorkspaceUpdate, session: Session = Depends(get_session)):
    return workspace_service.update_workspace(session, workspace_id, workspace_in)

@router.delete("/{workspace_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_workspace(workspace_id: str, session: Session = Depends(get_session)):
    workspace_service.delete_workspace(session, workspace_id)

File: frontend/src/lib/api/workspaces.ts
import { fetchApi } from '../api-client';

export interface Workspace {
  id: string;
  name: string;
  description: string | null;
  created_at: string;
  updated_at: string;
  dataset_count: number;
}

export interface WorkspaceCreate {
  name: string;
  description?: string;
}

export interface WorkspaceUpdate {
  name?: string;
  description?: string;
}

export const workspacesApi = {
  getWorkspaces: () => fetchApi<Workspace[]>('/workspaces'),
  getWorkspace: (id: string) => fetchApi<Workspace>(`/workspaces/${id}`),
  createWorkspace: (data: WorkspaceCreate) => fetchApi<Workspace>('/workspaces', {
    method: 'POST',
    body: JSON.stringify(data),
  }),
  updateWorkspace: (id: string, data: WorkspaceUpdate) => fetchApi<Workspace>(`/workspaces/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  }),
  deleteWorkspace: (id: string) => fetchApi<void>(`/workspaces/${id}`, {
    method: 'DELETE',
  }),
};
